[WIP] Update `LoraConfig` for KaSA implementation

nsbg opened the pull request:
cc @BenjaminBossan 

I was delayed in updating the code because I was focusing on company work, but now I'm planning to resume the project in earnest. If I have any questions about implementing the code, may I continue to ask you?

I apologize for opening a new pull request, as the previous one was closed 🥲 Thank you for your understanding.

BenjaminBossan reviewed (changes requested):
Thank you for resuming your work on KaSA.

Implementation-wise, we need to take a different approach. Right now, KaSA is just added to the normal LoRA code, but we only want to activate it if the user opts in. Therefore, it should be implemented in a separate class, something like `KasaVariant`, in `peft/tuners/lora/variants.py`. Please check how DoRA is implemented and use a similar approach, as I have detailed in [my previous comment](https://github.com/huggingface/peft/pull/2543#issuecomment-2930428128). If anything is unclear, feel free to ask.

BenjaminBossan commented:
gentle ping @nsbg

nsbg commented:
Thank you for your alert!

I spent some time looking over the KaSA paper and code to get ready for more serious work, but it does seem pretty difficult 🥲 My goal is to upload code that's ready for review before the end of September, so I'm going to try even harder.

Right now, I'm stuck at the 'Extend LoRA variant resolution' stage you mentioned. Honestly, this seems like the most important part, but it's hard for me to figure out where to start—specifically, which file and class I should work on first. Could you help me with this?

BenjaminBossan reviewed (changes requested):
Thanks for integrating my feedback. I gave this another review and noted the next few changes that are necessary. Please check my comments.

Apart from this, the branch is now encountering merge conflicts. Could you please bring your fork up-to-date with the remote and then merge with, or rebase on, the latest main branch from PEFT? If you have questions on how to resolve the merge conflicts, don't hesitate to ask.

Furthermore, please always run `make style` on your changes before pushing to make our linter happy.

More of a note for myself: Since KaSA updates the base weights of the model, we will have to take extra care to ensure that it works correctly when saving and loading the adapter.

nsbg commented on src/peft/tuners/lora/variants.py:
```diff
@@ -308,3 +309,56 @@ class DoraConv3dVariant(_DoraConvNdVariant):
+        # initialize lora_diag
+        module.lora_diag[adapter_name] = nn.Parameter(torch.randn(module.r[adapter_name]), requires_grad=True)
+
+        # SVD
```
```
# initialize lora_diag
module.lora_diag[adapter_name] = nn.Parameter(torch.randn(module.r[adapter_name]), requires_grad=True)

# see https://github.com/juyongjiang/KaSA/blob/f85e88c22d0fa4cb8ab2923d7c2bf1bbec152da3/peft/src/peft/tuners/lora/layer.py#L132
        
# SVD
```
I put it in here, how is it?

nsbg commented:
@BenjaminBossan oh I didn't mean to close the branch, but it seems to have closed while I was merging with the main branch. I guess I'll have to open a new PR, right? 😰

nsbg commented on src/peft/tuners/lora/variants.py:
```diff
@@ -308,3 +309,56 @@ class DoraConv3dVariant(_DoraConvNdVariant):
+    @staticmethod
+    def unmerge(module: Linear, active_adapter: str, orig_weight: torch.Tensor) -> torch.Tensor:
+        delta_weight = module.get_delta_weight(active_adapter)
+        return orig_weight - delta_weight
```
Although this PR is closed, it seems I've incorporated everything else except for this comment (of course, you'd have to look at the code). Could you explain this question in more detail?

BenjaminBossan commented:
> oh I didn't mean to close the branch, but it seems to have closed while I was merging with the main branch. I guess I'll have to open a new PR, right? 😰
> 
> +) when I tried to sync with the main branch, I ended up discarding all my commits, so did that cause it to close?

I don't know what happened, but I could re-open the PR and there are some changes visible. Can you double check that everything looks as expected? If for some reason it's not what it's expected, you can create a new PR and push your local branch.

nsbg commented:
I usually handle merges in the terminal, and I suspect the pull request was closed because I accidentally wiped the commit history while using the 'Sync fork' feature on GitHub. I'll be more careful in the future. Thanks for reopening it.

I'll review the changes and open a new PR if needed. Sorry to keep bothering you with this.

BenjaminBossan commented:
> I'll review the changes and open a new PR if needed. Sorry to keep bothering you with this.

No worries. If the diff on this PR looks good, let me know and I'll do a review. Only open a new PR if for some reason, the code here does not correspond to what it should be.

BenjaminBossan commented on src/peft/tuners/lora/variants.py:
```diff
@@ -438,6 +438,57 @@ def init(module: Conv3d, adapter_name: str, **kwargs: Any) -> None:
+        scaling = module.scaling[active_adapter]
+        diag = torch.diag(module.lora_diag[active_adapter])
+        
+        x = module._cast_input_dtype(x, lora_A.weight.dtype)
+        if isinstance(dropout, nn.Identity) or not module.training:
+            x = dropout(x)
+        
```
These lines can be removed, they're already being called by the normal `forward` of the LoRA layer.

BenjaminBossan reviewed (changes requested):
Thanks for the updates, we're making good progress here.

> The logic for raising errors in each layer hasn’t been applied yet, but I committed first to check whether adding the parameter in this way matches what you meant. Excluding the Linear class, it seems like an error should be raised when use_kasa is true in the other classes.

Nice, this looks correct, please raise the error in the unsupported layers as indicated.

> Also, I noticed there’s no part that calls KasaLinearVariant—should this be called inside the linear class? I’m a bit confused about this part.

Don't worry, it is being called. E.g. in `lora.Linear.forward` we have this code:

https://github.com/huggingface/peft/blob/190f9873b15660d9092f70065c18e4993fe10d5b/src/peft/tuners/lora/layer.py#L806-L816

So if the KaSA variant is found, `KasaLinearVariant.forward` will be used here. Same for the other methods.

nsbg commented:
@BenjaminBossan 
I removed unnecessary parts and added the `_get_delta_weight` static method specifically for the `KasaLinearVariants` class. If adding `_get_delta_weight` in this way is correct, I plan to update all occurrences of `module.get_delta_weight` within the current `KasaLinearVariants` class to use the modified method. Please review and share your thoughts.

BenjaminBossan commented on src/peft/tuners/lora/variants.py:
```diff
@@ -439,6 +439,64 @@ def init(module: Conv3d, adapter_name: str, **kwargs: Any) -> None:
+            
+    @staticmethod
+    def merge_safe(module: Linear, active_adapter: str, orig_weight: torch.Tensor) -> torch.Tensor:
+        delta_weight = module.get_delta_weight(active_adapter)
+        return orig_weight + delta_weight
+
+    @staticmethod
+    def merge_unsafe(module: Linear, active_adapter: str, orig_weight: torch.Tensor) -> None:
+        delta_weight = module.get_delta_weight(active_adapter)
+        orig_weight.data += delta_weight
+
+    @staticmethod
+    def unmerge(module: Linear, active_adapter: str, orig_weight: torch.Tensor) -> torch.Tensor:
+        delta_weight = module.get_delta_weight(active_adapter)
```
Instead of `module.get_delta_weight`, these methods should call the newly added `KasaLinearVariant._get_delta_weight`, right?

nsbg commented:
I've made the changes you commented on and added the following to the test file.  I used the name 'Vanilla MLP LoRA with KaSA' for the lines added to the test file, which is different from what you suggested. Please let me know if this wasn't your intention.

Also, I have a  question: in forward method, 
```@staticmethod
def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.Tensor) -> torch.Tensor:
  lora_A = module.lora_A[active_adapter]
  lora_B = module.lora_B[active_adapter]
  dropout = module.lora_dropout[active_adapter]
  scaling = module.scaling[active_adapter]
  diag = torch.diag(module.lora_diag[active_adapter])

  # KASA calculation
  # see https://github.com/juyongjiang/KaSA/blob/f85e88c22d0fa4cb8ab2923d7c2bf1bbec152da3/peft/src/peft/tuners/lora/layer.py#L602C21-L602C110
  lora_output = lora_B(torch.einsum("ijk,kl->ijl", lora_A(x), diag)) * scaling
  return result + lora_output
```
`dropout` variable isn't being used. Should I remove it or just leave it as is?

BenjaminBossan commented:
> `dropout` variable isn't being used. Should I remove it or just leave it as is?

Good catch. I think the correct way is to apply it to `x`, like in the original code: https://github.com/juyongjiang/KaSA/blob/f85e88c22d0fa4cb8ab2923d7c2bf1bbec152da3/peft/src/peft/tuners/lora/layer.py#L602C21-L602C110

nsbg commented:
Sorry for the delayed update; I've been working on this in bits and pieces due to personal scheduling. Since some time has passed, I'll summarize the feedback you previously gave me and my related work:

## 1. `_skip_test_disable_adapters` function
- Added skip logic for test functions that do not work with KaSA adapters. (you can check in `tests/test_custom_models.py`)
## 2. `TestKasaInitialization` class
- `test_kasa_mixed_adapter_error` method: When existing adapters are not KaSA, adding a new KaSA adapter should raise a ValueError.
- `test_kasa_mixed_adapter_error_reverse` method: When existing adapters are KaSA, adding a non-KaSA adapter should raise a ValueError.

Both of these test functions passed because they successfully raised a ValueError. My understanding is that "PASSED" here means the ValueError was raised, which is the intended behavior for these functions. Please correct me if my understanding is wrong.
## 3. Logic to avoid repetitive SVD processes (Just an idea 💡)
I agree with your opinion that repeatedly modifying the base weights for the same KaSA adapter is not ideal. My current idea is to apply a cache, as shown below, and check if SVD has already been applied by checking the existence of this attribute. What do you think of this approach?
```python
if not hasattr(module, "_kasa_svd_cache"):
    # First KaSA adapter: perform SVD and cache the result
    weight = module.get_base_layer().weight
    dtype = weight.dtype
    weight = weight.to(torch.float32)
    U, S, Vh = torch.linalg.svd(weight.data, full_matrices=False)
    module._kasa_svd_cache = (U, S, Vh)
else:
    # Reuse cached SVD result
    U, S, Vh = module._kasa_svd_cache
    dtype = module.get_base_layer().weight.dtype
```

BenjaminBossan commented on tests/test_initialization.py:
```diff
@@ -4477,3 +4477,67 @@ def test_key_mapping_save_old_load_new_vblora(self, old_model, new_model, tmp_pa
         self.check_vblora_load_no_warning(new_model, old_model, tmp_path)
+
+
+class TestKasaInitialization:
```
Could you please move the test functions to the bottom of `class TestLoraInitialization`? This is because KaSA is a LoRA variant, and it also means we can reuse the existing `get_model` method.

BenjaminBossan commented on tests/test_initialization.py:
```diff
@@ -4477,3 +4477,67 @@ def test_key_mapping_save_old_load_new_vblora(self, old_model, new_model, tmp_pa
+        return MyModule().eval().to(self.torch_device)
+
+    @pytest.fixture
+    def data(self):
```
This fixture is not needed, as `data` is not used in the tests.

BenjaminBossan reviewed (changes requested):
Thanks for the updates, we're moving towards the goal. I commented on some of the changes, please check.

> My current idea is to apply a cache, as shown below, and check if SVD has already been applied by checking the existence of this attribute. What do you think of this approach?

I'm not sure it's needed, but I could be wrong. AFAICT, the SVD is only needed to update the base weight. If we add a second KaSA adapter, the base weight is already updated, so there is no need to cache `U, S, Vh`, we can just skip completely. Is my understanding correct?

Also, there are some merge conflicts on this PR now. I think they should be easy to resolve, but don't hesitate to ask if you have questions. Finally, before committing, don't forget to call `make style`.

BenjaminBossan commented on src/peft/tuners/lora/model.py:
```diff
@@ -155,6 +155,29 @@ class LoraModel(BaseTuner):
+            raise ValueError(
+                f"{self.__class__.__name__} supports only 1 adapter with bias. When using multiple adapters, "
+                "set bias to 'none' for all adapters."
+            )
```
Let's remove this and call `super()._check_new_adapter_config(config)` instead.

iambogeumkim commented:
I did run `make style`, but it looks like one file was missed. I re-ran it and pushed the commit.

Also, thank you for your patience with all my questions, even the trivial ones. I know I might have been a bit of a bother 😅 Since this was my first code contribution, I learned so much thanks to your guidance. Wishing you a warm and happy holiday season!

githubnemo commented:
*not stale*
